File: cpp/src/2023/day_15.cpp
#include "../utils/input.hpp"
#include <cassert>

auto hash(const std::string &s) {
  int result{0};
  for (const auto c : s) {
    result += int(c);
    result *= 17;
    result %= 256;
    // std::cout << std::format("Hashing: {} -> {}", s, c, result) << std::endl;
  }
  return result;
}

auto part_one(const std::vector<std::string> &input) {
  int result{0};
  for (const auto &line : input) {
    auto splits = utils::split_string(line, ',');
    for (const auto &split : splits) {
      result += hash(split);
    }
  }
  return result;
}

auto part_two(const std::vector<std::string> &input) { return 2; }

int main() {
  // std::filesystem::path input_path{"../../data/2023/input_15_mock.txt"};
  std::filesystem::path input_path{"../../data/2023/input_15.txt"};
  utils::Reader reader(input_path);
  auto input = reader.get_lines();

  std::cout << std::format("The answer to part one is: {}", part_one(input))
            << std::endl;
  std::cout << std::format("The answer to part two is: {}", part_two(input))
            << std::endl;

  return 0;
}

File: python/2023/day_15.py
from collections import defaultdict
from pathlib import Path


def parse(input_str: str) -> list[str]:
    return input_str.strip().split(",")


def hash_value(s: str) -> int:
    current = 0
    for char in s:
        current += ord(char)
        current *= 17
        current %= 256
    return current


def part_one(steps: list[str]) -> int:
    return sum(hash_value(step) for step in steps)


def part_two(steps: list[str]) -> int:
    boxes = defaultdict(dict)

    for step in steps:
        if "-" in step:
            label = step[:-1]
            box_id = hash_value(label)
            if label in boxes[box_id]:
                del boxes[box_id][label]
        elif "=" in step:
            label, focal_length = step.split("=")
            box_id = hash_value(label)
            boxes[box_id][label] = int(focal_length)

    total = 0
    for box_id, lenses in boxes.items():
        for slot, (label, focal_length) in enumerate(lenses.items(), 1):
            focusing_power = (box_id + 1) * slot * focal_length
            total += focusing_power

    return total


if __name__ == "__main__":
    input_path = Path(__file__).parent.parent.parent / "data/2023/input_15.txt"
    with open(input_path, "r") as f:
        input_str = f.read()
        steps = parse(input_str)

    print("Part One:", part_one(steps))
    print("Part Two:", part_two(steps))
